lib: add aws request signing library

efd6 opened the pull request:
This adds Go implementations of v4 AWS request signing functions to ease their use.

For elastic/beats#40762

Please take a look.

kcreddy commented on lib/aws.go:
```diff
@@ -0,0 +1,293 @@
+	if !ok {
+		return types.ValOrErr(args[2], "no such overload for "+name)
+	}
+	token, ok := args[3].(types.String)
+	if !ok {
+		return types.ValOrErr(args[2], "no such overload for "+name)
+	}
```
```suggestion
	token, ok := args[3].(types.String)
	if !ok {
		return types.ValOrErr(args[3], "no such overload for "+name)
	}
```

kcreddy commented on lib/aws.go:
```diff
@@ -0,0 +1,293 @@
+	}
+	creds := credentials.NewStaticCredentialsFromCreds(credentials.Value{
+		AccessKeyID:     string(id),
+		SecretAccessKey: string(token),
```
```suggestion
		SecretAccessKey: string(secret),
```

kcreddy commented on lib/aws.go:
```diff
@@ -0,0 +1,293 @@
+// This corresponds to using [v4_creds.Signer.Sign] using the result of a call
+// to [credentials.NewSharedCredentials].
+//
+// # Sign AWS from shared credentials
```
```suggestion
// # Sign AWS from static credentials
```

kcreddy commented on lib/aws.go:
```diff
@@ -0,0 +1,293 @@
+// Examples:
+//
+//	req.sign_aws_from_static("id", "secret", "token", "service", "region", now(), false, false, false)   // return {"Body": "…
+//	req.sign_aws_from_static("id", "secret", "token", "provider", "service", "region", now(), false, false, false)   // return {"Body": "…
```
```suggestion
//	req.sign_aws_from_static("access_key_id", "secret_access_key", "session_token", "provider", "service", "region", now(), false, false, false)   // return {"Body": "…
```
Maybe easy to comprehend 🤔. I will leave to you.

efd6 replied:
I think I'm OK with what is there, and the signature is already too long, so I'd rather keep it as is.

efd6 commented:
The snyk complaint here is https://security.snyk.io/vuln/SNYK-GOLANG-GITHUBCOMAWSAWSSDKGOSERVICES3S3CRYPTO-598253. This is not valid since that code is not called by us.

govulncheck output:
```
Your code is affected by 3 vulnerabilities from the Go standard library.
This scan also found 0 vulnerabilities in packages you import and 2
vulnerabilities in modules you require, but your code doesn't appear to call
these vulnerabilities.
```
The three vulnerabilities show up because the go.mod go directive allows unfixed 1.24 versions (below go1.24.4). This is a consequence of how version constraints are handled by go.mod; it is not possible to express 1.23.10 or 1.24.4.

The snyk complaints are the following module issues, mentioned in the govulncheck output, that we do not call.
```
=== Module Results ===

Vulnerability #1: GO-2022-0646
    CBC padding oracle issue in AWS S3 Crypto SDK for golang in
    github.com/aws/aws-sdk-go
  More info: https://pkg.go.dev/vuln/GO-2022-0646
  Module: github.com/aws/aws-sdk-go
    Found in: github.com/aws/aws-sdk-go@v1.55.7
    Fixed in: N/A

Vulnerability #2: GO-2022-0635
    In-band key negotiation issue in AWS S3 Crypto SDK for golang in
    github.com/aws/aws-sdk-go
  More info: https://pkg.go.dev/vuln/GO-2022-0635
  Module: github.com/aws/aws-sdk-go
    Found in: github.com/aws/aws-sdk-go@v1.55.7
    Fixed in: N/A
```
